### PR TITLE
optional config to apply blockTimeout and gasFeeMultiplier

### PR DESCRIPTION
**Pull request by soonhuat**
Fixes # .

Changes proposed in this PR:

- add optional config parameter when init contract
- add optional config for token approve
- both allow market operator to set different transaction and multiple config

### CODE REVIEW

**Comment by alexcos20** [comment]
@bogdanfazakas - is this what you need ?

**Comment by bogdanfazakas** [comment]
> @bogdanfazakas - is this what you need ?

Will check

File: src/pools/ssContracts/SideStaking.ts [hunk 1]
```diff
@@ -350,7 +369,11 @@ export class SideStaking {
     contractInstance?: Contract
   ): Promise<number> {
     const sideStaking =
-      contractInstance || new this.web3.eth.Contract(this.ssAbi as AbiItem[], ssAddress)
+      contractInstance ||
+      setContractDefaults(
```
Comment (bogdanfazakas) [inline src/pools/ssContracts/SideStaking.ts:373]: I think you can use also here the newly added `sideStakingContract()` method

File: src/pools/ssContracts/SideStaking.ts [hunk 2]
```diff
@@ -286,7 +301,11 @@ export class SideStaking {
...
```
Comment (bogdanfazakas) [inline src/pools/ssContracts/SideStaking.ts:305]: same here as prev comment ☝️

File: src/pools/ssContracts/SideStaking.ts [hunk 3]
```diff
@@ -21,11 +26,18 @@ export class SideStaking {
...
   }
 
+  private sideStakingContract(ssAddress: string) {
+    return setContractDefaults(
+      new this.web3.eth.Contract(this.ssAbi, ssAddress),
+      this.config
+    )
+  }
+
   async amountToUnits(token: string, amount: string): Promise<string> {
```
Comment (kremalicious) [inline src/pools/ssContracts/SideStaking.ts:36]: there are 3 places in the codebase with exactly the same code. Why? So if we do a change here do we need to change all other places too?

Screen Shot 2022-04-14 at 16 45 13

File: src/pools/ssContracts/SideStaking.ts [hunk 4]
```diff
@@ -40,9 +52,9 @@ export class SideStaking {
 
   async unitsToAmount(token: string, amount: string): Promise<string> {
```
Comment (kremalicious) [inline src/pools/ssContracts/SideStaking.ts:53]: same here

Screen Shot 2022-04-14 at 16 46 00

File: src/pools/ssContracts/SideStaking.ts (hunk 3 above)
Comment (kremalicious) [reply src/pools/ssContracts/SideStaking.ts:36]: not fully same code but this is just crazy confusing and there's no way for reviewer to know what the intention here was or why there are methods with exact same name in multiple files.

In any case this is then also populated to CodeClimate:

Screen Shot 2022-04-14 at 16 48 06

File: src/utils/TokenUtils.ts [hunk 5]
```diff
@@ -47,18 +53,23 @@ export async function estApprove(
...
+  config?: Config
 ): Promise<TransactionReceipt | string> {
-  const tokenContract = new web3.eth.Contract(minAbi, tokenAddress)
+  const tokenContract = setContractDefaults(
+    new web3.eth.Contract(minAbi, tokenAddress),
+    config
+  )
   if (!force) {
-    const currentAllowence = await allowance(web3, tokenAddress, account, spender)
+    const currentAllowence = await allowance(web3, tokenAddress, account, spender, config)
```
Comment (kremalicious) [inline src/utils/TokenUtils.ts:72]: typo:

```suggestion
    const currentAllowance = await allowance(web3, tokenAddress, account, spender, config)
```

Comment (kremalicious) [inline src/utils/TokenUtils.ts:72]: typo:

```suggestion
    const currentAllowance = await allowance(web3, tokenAddress, account, spender, config)
```

**Comment by soonhuat** [comment]
> No writeup of problem or intention so hard to review fully. Just from DX perspective it feels weird having to construct a full `Config` object to get like 1-3 values across. Sounds like this should have its own interface of values which actually influence anything in the `setContractDefaults` which seems to be just reassigning values. So looks like a deeper architectural problem and I guess was not intention of this PR to solve, so we just continue to hack in stuff to solve isolated issues. But for developer completely unclear what exact values the full `Config` then should have if I just want to change the `gasFeeMultiplier`.
> 
> Also, is this intended to be merged before contracts 1.0.0-alpha.29? Cause that version then has no vesting anymore but this PR here touches multiple vesting methods

@kremalicious  @bogdanfazakas 
wonder does this look good?
the vesting part, i just apply the same implementation, when alpha-29 in, will auto remove the area of changes right. would be fine either before or after alpha-29 merge in.

as for code climate, looks like mostly is the warning of > 4 parameter, and false positive of redundant code duplication.

**Comment by AnaLoznianu** [comment]
@soonhuat The AMM pool removal PR: https://github.com/oceanprotocol/ocean.js/pull/1568 adds a lot of changes to the ocean.js lib. Once @bogdanfazakas merges that PR, please update this one if it is still relevant. Thank you!

**Comment by soonhuat** [comment]
> @soonhuat The AMM pool removal PR: #1568 adds a lot of changes to the ocean.js lib. Once @bogdanfazakas merges that PR, please update this one if it is still relevant. Thank you!

@AnaLoznianu could abandon this PR, I created another PR from fork on top of eip-1559 working on by Bogdan
https://github.com/oceanprotocol/ocean.js/pull/1581

**Comment by AnaLoznianu** [comment]
All right, thank you for the update. Closing this then

**Pull request closed without merging**